phpBB/phpbb/ban/type/ip.php

```diff
@@ -0,0 +1,524 @@
+<?php
+/**
+ *
+ * This file is part of the phpBB Forum Software package.
+ *
+ * @copyright (c) phpBB Limited <https://www.phpbb.com>
+ * @license GNU General Public License, version 2 (GPL-2.0)
+ *
+ * For full copyright and license information, please see
+ * the docs/CREDITS.txt file.
+ *
+ */
+
+namespace phpbb\ban\type;
+
+use Symfony\Component\HttpFoundation\IpUtils;
+
+class ip extends base
+{
+	/**
+	 * {@inheritdoc}
+	 */
+	public function exclude_possible()
+	{
+		return false;
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function get_type()
+	{
+		return 'ip';
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function get_user_column()
+	{
+		return 'user_ip';
+	}
+
+	/**
+	 * {@inheritdoc}
+	 */
+	public function add_ban(array $ban_list, $ban_end, $ban_exclude, $ban_reason, $ban_reason_display)
+	{
+		$banned_ips = array();
+
+		foreach ($ban_list as $ban_entry)
+		{
+			$ban_entry = trim($ban_entry);
+
+			// Attention, we're banning a range of IPs
+			if (stripos($ban_entry, '-') !== false)
+			{
+				// Check 'em
+				$ips = array_map('trim', explode('-', $ban_entry));
+				if ((!preg_match(get_preg_expression('ipv4'), $ips[0]) || !preg_match(get_preg_expression('ipv4'), $ips[1]))
+				&& (!preg_match(get_preg_expression('ipv6'), $ips[0]) || !preg_match(get_preg_expression('ipv6'), $ips[1])))
+				{
+					continue;
+				}
+
+				// We will convert it later, just add it for now
+				$banned_ips[] = implode('-', $ips);
+			}
+			// we will check it for validity better later
+			else if (preg_match('#^\d{1,3}(?:\.[0-9*]{1,3}){1,3}(?:/\d{1,2})?$#', $ban_entry))
+			{
+				$ban_item = array_map('trim', explode('/', $ban_entry, 2));
+
+				$octets = substr_count($ban_item[0], '.') + 1;
+
+				if ($octets > 4)
+				{
+					continue;
+				}
+				else if ($octets < 4)
+				{
+					$ban_item[0] .= str_repeat('.*', 4 - $octets);
+				}
+
+				// Do a dirty trick to check if it's valid:
+				if (!preg_match(get_preg_expression('ipv4'), str_replace('*', '255', $ban_item[0])))
+				{
+					continue;
+				}
+
+				if (sizeof($ban_item == 2))
+				{
+					if ($ban_item[1] > 32)
+					{
+						continue;
+					}
+
+					$ban_item[0] = str_replace('*', '0', $ban_item[0]);
+				}
+
+				//$wildcards = substr_count($ban_entry, '*');
+				//$banned_ips[] = str_replace('*', '0', $ban_entry) . '/' . (32 - $wildcards * 4);
+				$banned_ip = implode('/', $ban_item);
+				if (stripos($banned_ip, '/') !== false)
+				{
+					$banned_ip = $this->normalise_cidr($banned_ip);
+				}
+				$banned_ips[] = $banned_ip;
+			}
+			// Okay, we are really lazy here...
```

Comment:
I am sorry. :D But the ban function is not in its final state - I will probably change everything again. This applies to everything I already wrote because I'm almost never satisfied with the result, I have.
However, I will look through everything and change the comments if necessary.